File: solutions/fsharp/robot-simulator/1/RobotSimulator.fs
module RobotSimulator

type Direction =
    | North
    | East
    | South
    | West

type Position = int * int

type Robot =
    { Direction: Direction
      Position: Position }

let private directionToInt direction =
    match direction with
    | North -> 0
    | East -> 1
    | South -> 2
    | West -> 3

let private intToDirection i =
    match i with
    | 0 -> North
    | 1 -> East
    | 2 -> South
    | _ -> West

let private changeDir direction mov =
    let directionInt = directionToInt direction
    let newDirectionInt = (directionInt + (mov + 4)) % 4
    intToDirection newDirectionInt

let create direction position =
    { Direction = direction
      Position = position }

let private updatePosition (x, y) direction =
    match direction with
    | North -> (x, y + 1)
    | East -> (x + 1, y)
    | South -> (x, y - 1)
    | West -> (x - 1, y)

let rec moveHelper (instructions: char list) robot =
    match instructions with
    | [] -> robot
    | c :: rest ->
        let newRobot =
            match c with
            | 'R' ->
                { robot with
                    Direction = changeDir robot.Direction 1 }
            | 'L' ->
                { robot with
                    Direction = changeDir robot.Direction -1 }
            | _ ->
                { robot with
                    Position = updatePosition robot.Position robot.Direction }

        moveHelper rest newRobot

let move (instructions: string) robot =
    moveHelper (instructions |> List.ofSeq) robot

File: solutions/python/robot-simulator/1/robot_simulator.py
EAST = 0
NORTH = 1
WEST = 2
SOUTH = 3
_DIR_END = 4


class Robot:
    _advance = {
        EAST: (lambda x, y: (x + 1, y)),
        NORTH: (lambda x, y: (x, y + 1)),
        WEST: (lambda x, y: (x - 1, y)),
        SOUTH: (lambda x, y: (x, y - 1)),
    }

    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, moves):
        for m in moves:
            if m == "R":
                self.direction = (self.direction - 1 + _DIR_END) % _DIR_END
            elif m == "L":
                self.direction = (self.direction + 1) % _DIR_END
            else:  # assume m == "A"
                self.coordinates = Robot._advance[self.direction](*self.coordinates)
